js/DataService.js

```diff
@@ -8,16 +8,98 @@ angular.module('PathOfDamage')
     var int = parseInt(string);
     return isNaN(int) ? null : int;
   }
+  
+  function Row(enabled, name, value, elements) {
+    this.enabled = enabled || true;
+    this.name = name || '';
+    this.value = value || null;
+    this.elements = elements;
+    this.isValid = function () {
+      return this.enabled && this.value;
+    }
+  }
 
-  function checkTotal(table) {
-    if (table.totalMin) {
-      table.total = Math.max(table.total, table.totalMin);
+  function Table(name, addToTotal, defaultElements) {
+    this.name = name;
+    this.defaultElements = defaultElements;
+    this.rows = [];
+    this.calcTotals = function () {
+      this.totals = this.getEmptyTotals();
+      for (var i = 0, len = this.rows.length - 1; i < len; i++) {
+        if (this.rows[i].isValid()) {
+          addToTotal(this.totals, this.rows[i]);
+        }
+      }
+    };
+    this.addRow = function (enabled, name, value, elements) {
+      this.rows.push(new Row(enabled, name, value, elements || this.getDefaultElements()));
+    };
+    this.getDefaultElements = function () {
+      return this.defaultElements ? defaultElements.slice() : undefined;
```

Comment:
I'm not thrilled about this, but angular doesn't know to copy the array since ng-model is item 0 of the array, so I have to copy it manually here when it's fetched for creating rows, otherwise all the rows are linked for damage shifts (other tables are fine)